pysal/contrib/viz/mapping.py

```diff
@@ -610,11 +611,23 @@ def geoplot(db, col=None, palette='BuGn', classi='Quantiles',
                   Additional named vaues to be passed to the classifier of choice.
     '''
     if col:
+        if hasattr(palette, 'number') and 'k' not in kwargs:
+            # infer number of classes from available colors
+            kwargs['k'] = palette.number
+        elif hasattr(palette, 'number') and 'k' in kwargs:
+            if kwargs['k'] != palette.number:
+                warn('explicit "k" argument differs from the number of colors in '
+                'the palette. Defaulting to the number of colors in the palette.', 
+                stacklevel=1)
+                kwargs['k'] = palette.number
```

Comment:
This branch should warn the user about that disagreement now and select `palette.number` over `k` if they disagree. 

It'd be nice if `k` would supersede `palette.number`, so that if you wanted just the first three from a 5-color categorical palette, you could do so. But, that's not really a big deal I think.